sdk/core/core-util/src/aborterUtils.ts

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortSignalLike, AbortController } from "@azure/abort-controller";
+
+/**
+ * Options related to abort controller.
+ */
+export interface AbortOptions {
+  /**
+   * The abortSignal associated with containing operation.
+   */
+  abortSignal?: AbortSignalLike;
+  /**
+   * The abort error message associated with containing operation.
+   */
+  abortErrorMsg?: string;
+}
+
+/**
+ * Represents a function that returns a promise that can be aborted.
+ */
+export type AbortablePromiseBuilder<T> = (abortOptions: {
+  abortSignal?: AbortSignalLike;
+}) => Promise<T>;
+
+/**
+ * promise.race() wrapper that aborts rest of promises as soon as the first promise settles.
+ */
+export async function cancelablePromiseRace<T extends unknown[]>(
+  abortablePromiseBuilders: AbortablePromiseBuilder<T[number]>[],
+  options?: { abortSignal?: AbortSignalLike }
+): Promise<T[number]> {
+  const aborter = new AbortController();
+  options?.abortSignal?.addEventListener("abort", () => aborter.abort());
+  try {
+    return await Promise.race(
+      abortablePromiseBuilders.map((p) => p({ abortSignal: aborter.signal }))
+    );
+  } finally {
+    aborter.abort();
+    options?.abortSignal?.removeEventListener("abort", () => aborter.abort());
```

Comment:
I don't think this will work since the anonymous function `() => aborter.abort()` will have a different identity than the one registered above. I believe you need to pull it out into a local like

```ts
function abortHandler() {
    aborter.abort();
}
options?.abortSignal?.addEventListener("abort", abortHandler);
try {
// etc
} finally {
options?.abortSignal?.removeEventListener("abort", abortHandler);
}
```

in the future we can use addEventListener with the 'once' option to avoid this dance, but the above will work everywhere.

Maybe this event handler not being unregistered is the source of your small leak?

Reply:
Ah, a great catch, will check that.